client/cli/src/commands/export_sync_state_cmd.rs

```diff
@@ -0,0 +1,96 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::error;
+use crate::params::{SharedParams, NetworkParams};
+use crate::CliConfiguration;
+use crate::Role;
+use log::info;
+use sc_service::ChainSpec;
+use sc_service::chain_ops::{MaybeChtRootStorageProvider, build_light_sync_state};
+use structopt::StructOpt;
+use std::io::Write;
+use std::sync::Arc;
+use std::task::Poll;
+use sp_runtime::traits::Block as BlockT;
+
+/// Export the chain spec containing a state to sync the light client.
+#[derive(Debug, StructOpt)]
+pub struct ExportSyncStateCmd {
+	/// Force raw genesis storage output in the chain spec.
+	#[structopt(long = "raw")]
+	pub raw: bool,
+	
+	#[allow(missing_docs)]
+	#[structopt(flatten)]
+	pub shared_params: SharedParams,
+
+	#[allow(missing_docs)]
+	#[structopt(flatten)]
+	pub network_params: NetworkParams,
+}
+
+impl ExportSyncStateCmd {
+	/// Run the `export-sync-state` command
+	pub async fn run<B, CL, BA, SO>(
+		&self,
+		mut spec: Box<dyn ChainSpec>,
+		client: Arc<CL>,
+		backend: Arc<BA>,
+		mut sync_oracle: SO,
+	) -> error::Result<()>
+		where
+			B: BlockT,
+			CL: sp_blockchain::HeaderBackend<B>,
+			BA: MaybeChtRootStorageProvider<B>,
+			SO: sp_consensus::SyncOracle,
+	{
+		futures::future::poll_fn(|_| {
+			match sync_oracle.is_major_syncing() || sync_oracle.is_offline() {
+				true => Poll::Pending,
+				false => Poll::Ready(())
+			}
+		}).await;
```

Comment:
ooh this is pretty clever :grinning:, shouldn't you call the `ctx.wake()` somewhere in the fn though?

Reply:
Not sure. @tomaka should know.

Reply:
The code you have written is essentially just:

```
while sync_oracle.is_major_syncing() || sync_oracle.is_offline() {}
```

Reply:
Ahh no, not even that. It just calls this future once and than it is stuck, because nothing wakes the future again.

Reply:
Yeah, I believe you should call `client.import_notification_stream()` and check `is_major_sync() || is_offline()` in response to each newly imported block.
Maybe add a timeout in case we're indeed offline.

Reply:
Using `NetworkStatusSinks` would actually be a better idea, no?

Reply:
It's not really meant for this kind of usage. The entire `NetworkStatus` is a hack.

Reply:
It is a... task that wakes up all the time and check the status of something? Why would anyone need something like that. At this rate you can just make a loop with a `sleep(1sec).await` in it :thinking: probably more readable too